### PR TITLE
Pytorch quantizers changes

### PR DESCRIPTION
**Pull request by reuvenperetz**
Add quantizers implementation to support ONNX export:
1. Add autograd function implementation for each quantizer to implement the op symbolic and forward static functions.
2. Add np implementation to support ops onnxruntime inference.

Change quantizers API to receive lists as their quantization range (instead of np arrays).

### CODE REVIEW

File: mct_quantizers/__init__.py [hunk 1]
```diff
@@ -26,4 +26,6 @@
 from mct_quantizers.keras import quantizers as keras_quantizers
 from mct_quantizers.pytorch import quantizers as pytorch_quantizers
 
+from mct_quantizers.pytorch.onnxruntime_session_options import get_ort_session_options
+
 __version__ = "1.2.0"
```
Comment (Idan-BenAmi) [inline mct_quantizers/__init__.py:31]: test

File: mct_quantizers/pytorch/quantizers/weights_inferable_quantizers/weights_uniform_inferable_quantizer.py [hunk 2]
```diff
@@ -97,6 +215,70 @@ def __call__(self,
...
+                num_bits: The number of bits to represent the quantized value.
+                min_range: min quantization range for quantizing weights
+                max_range: max quantization range for quantizing weights
+                per_channel: whether to use per-channel quantization
+                channel_axis: Axis of input to apply per-channel quantization on.
+
+            Returns:
+                The node in the ONNX graph representing the output of this operation.
+            """
+            return g.op("ai.onnx.contrib::WeightsUniformQuantizer", input_tensor,
```
Comment (Chizkiyahu) [inline mct_quantizers/pytorch/quantizers/weights_inferable_quantizers/weights_uniform_inferable_quantizer.py:263]: consider to use "mct_quantizers" in the domain

File: mct_quantizers/pytorch/onnxruntime_session_options.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+from mct_quantizers.common.constants import FOUND_ONNXRUNTIME, FOUND_ONNXRUNTIME_EXTENSIONS
+
+if FOUND_ONNXRUNTIME and FOUND_ONNXRUNTIME_EXTENSIONS:
+    import onnxruntime as ort
+    from onnxruntime_extensions import get_library_path
+
+    def get_ort_session_options() -> ort.SessionOptions:
+        """
+        :return: Session options for loading onnxruntime inference
+        session with custom implementation of onnx ops.
```
Comment (Idan-BenAmi) [inline mct_quantizers/pytorch/onnxruntime_session_options.py:25]: typo: Returns

File: mct_quantizers/pytorch/quantizers/activation_inferable_quantizers/activation_symmetric_inferable_quantizer.py [hunk 4]
```diff
@@ -12,17 +12,93 @@
...
+            min, max = 0, threshold - scale
+
+        quantized = np.round(np.clip(input_tensor, min, max) / scale) * scale
+        return quantized
+
 if FOUND_TORCH:
     import torch
     from mct_quantizers.pytorch.quantizers.base_symmetric_inferable_quantizer import BaseSymmetricInferableQuantizer
 
+    def quantize_sym_activations_torch(input_tensor: torch.Tensor,
```
Comment (Idan-BenAmi) [inline mct_quantizers/pytorch/quantizers/activation_inferable_quantizers/activation_symmetric_inferable_quantizer.py:75]: why not to use torch function: fake_quantize_per_tensor_affine

File: mct_quantizers/pytorch/quantizers/activation_inferable_quantizers/activation_symmetric_inferable_quantizer.py [hunk 5]
```diff
@@ -12,17 +12,93 @@
...
+                           signed,
+                           num_bits):
+        return quantize_sym_activations_numpy(input_tensor,
+                                              threshold,
+                                              signed,
+                                              num_bits)
+
+
+    def quantize_sym_activations_numpy(input_tensor: np.ndarray,
+                                       threshold: float,
```
Comment (Idan-BenAmi) [inline mct_quantizers/pytorch/quantizers/activation_inferable_quantizers/activation_symmetric_inferable_quantizer.py:45]: Confirm it is not already defined (maybe we can reuse existing function)

File: mct_quantizers/pytorch/quantizers/activation_inferable_quantizers/activation_uniform_inferable_quantizer.py [hunk 6]
```diff
@@ -12,15 +12,103 @@
...
+
 
 if FOUND_TORCH:
     import torch
     from mct_quantizers.pytorch.quantizers.base_uniform_inferable_quantizer import BaseUniformInferableQuantizer
+    from mct_quantizers.pytorch.quantizer_utils import fix_range_to_include_zero
+
+
+    def quantize_uniform_activations_torch(tensor_data: torch.Tensor,
+                                           range_min: float,
```
Comment (Idan-BenAmi) [inline mct_quantizers/pytorch/quantizers/activation_inferable_quantizers/activation_uniform_inferable_quantizer.py:81]: Same as above, check that we can't reuse existing function.

**Review by ofirgo — COMMENTED** [review]
@reuvenperetz 
please talk to me before you merge this PR, I want to go over the code with you to understand the changes

File: mct_quantizers/pytorch/quantizers/weights_inferable_quantizers/weights_pot_inferable_quantizer.py [hunk 7]
```diff
@@ -52,10 +77,95 @@ def __init__(self,
...
+            self.channel_axis = channel_axis
 
-            is_threshold_pot = np.all(np.round(np.log2(threshold.flatten()))==np.log2(threshold.flatten()))
+            is_threshold_pot = np.all(
+                np.round(np.log2(self.threshold_np.flatten())) == np.log2(self.threshold_np.flatten()))
             assert is_threshold_pot, f'Expected threshold to be power of 2 but is {threshold}'
 
+        def __call__(self, inputs: torch.Tensor) -> torch.Tensor:
+            """
+            Quantize the given inputs using the quantizer parameters.
```
Comment (Idan-BenAmi) [inline mct_quantizers/pytorch/quantizers/weights_inferable_quantizers/weights_pot_inferable_quantizer.py:91]: Quantize the given weights using the quantizer parameters.

File: mct_quantizers/pytorch/quantizers/weights_inferable_quantizers/weights_symmetric_inferable_quantizer.py [hunk 8]
```diff
@@ -12,19 +12,116 @@
...
+if FOUND_ONNXRUNTIME_EXTENSIONS:
+    from onnxruntime_extensions import onnx_op, PyCustomOpDef
+    def quantize_sym_weights_numpy(input_tensor: np.ndarray,
+                                   num_bits: int,
+                                   threshold: float,
+                                   per_channel: bool,
+                                   channel_axis: int):
+        """
+           Quantizes the input tensor symmetrically using numpy.
+
```
Comment (Idan-BenAmi) [inline mct_quantizers/pytorch/quantizers/weights_inferable_quantizers/weights_symmetric_inferable_quantizer.py:32]: Quantizes the weights tensor symmetrically using numpy

File: mct_quantizers/pytorch/quantizers/weights_inferable_quantizers/weights_uniform_inferable_quantizer.py [hunk 9]
```diff
@@ -12,19 +12,125 @@
...
+
+    def quantize_uniform_weights_numpy(input_tensor: np.ndarray,
+                                       num_bits: int,
+                                       min_range: np.ndarray,
+                                       max_range: np.ndarray,
+                                       per_channel: bool,
+                                       channel_axis: int):
+        """
+           Quantizes the input tensor symmetrically using numpy.
+
```
Comment (Idan-BenAmi) [inline mct_quantizers/pytorch/quantizers/weights_inferable_quantizers/weights_uniform_inferable_quantizer.py:37]: Same in here: replace input tensor with weights tensor,
Check if you can reuse existing quantizers function for weights

**Review by Idan-BenAmi — COMMENTED** [review]
No critical issue. I only suggest to check if we can reuse existing quantization functions instead to redefine them.

File: mct_quantizers/__init__.py (hunk 1 above)
Comment (reuvenperetz) [reply mct_quantizers/__init__.py]: It is tested in the tests since all tests are using it to check inference is possible

**Pull request merged**